lib/std/zon/parse.zig

```diff
@@ -0,0 +1,2832 @@
+                    try writer.writeAll(msg);
+                    inline for (info.fields, 0..) |field_info, i| {
+                        if (i != 0) try writer.writeAll(", ");
+                        try writer.print("{}", .{std.zig.fmtId(field_info.name)});
+                    }
+                    return self.failToken(.{
+                        .token = token,
+                        .offset = 0,
+                        .message = try buf.toOwnedSlice(self.gpa),
+                        .owned = true,
+                    });
+                }
+            },
+            else => @compileError("unreachable, should not be called for type " ++ @typeName(T)),
+        }
+    }
+
+    fn failExpectedContainer(self: @This(), T: type, node: Zoir.Node.Index) error{ OutOfMemory, ParseZon } {
+        @branchHint(.cold);
+        switch (@typeInfo(T)) {
+            .@"struct" => |@"struct"| if (@"struct".is_tuple) {
+                return self.failNode(node, "expected tuple");
+            } else {
+                return self.failNode(node, "expected struct");
+            },
+            .@"union" => return self.failNode(node, "expected union"),
+            .array => return self.failNode(node, "expected tuple"),
+            .pointer => |pointer| {
+                if (pointer.child == u8 and
+                    pointer.size == .slice and
+                    pointer.is_const and
+                    (pointer.sentinel_ptr == null or @as(*const u8, @ptrCast(pointer.sentinel_ptr)).* == 0) and
+                    pointer.alignment == 1)
+                {
+                    return self.failNode(node, "expected string");
+                } else {
+                    return self.failNode(node, "expected tuple");
+                }
+            },
+            else => {},
+        }
+        @compileError("unreachable, should not be called for type " ++ @typeName(T));
+    }
+
+    // Technically we could do this if we were willing to do a deep equal to verify
+    // the value matched, but doing so doesn't seem to support any real use cases
+    // so isn't worth the complexity at the moment.
+    fn failRuntimeValueComptimeVar(
+        self: @This(),
+        node: Zoir.Node.Index,
+        field: usize,
+    ) error{ OutOfMemory, ParseZon } {
+        @branchHint(.cold);
+        const ast_node = node.getAstNode(self.zoir);
+        var buf: [2]Ast.Node.Index = undefined;
+        const token = if (self.ast.fullStructInit(&buf, ast_node)) |struct_init| b: {
+            const field_node = struct_init.ast.fields[field];
+            break :b self.ast.firstToken(field_node);
+        } else b: {
+            const array_init = self.ast.fullArrayInit(&buf, ast_node).?;
+            const value_node = array_init.ast.elements[field];
+            break :b self.ast.firstToken(value_node);
+        };
+        return self.failTokenFmt(token, 0, "cannot store runtime value in compile time variable", .{});
```

Comment:
That's sort of a weird error here, because this is about runtime *fields*. How about: `"cannot initialize comptime field of 'Foo'"`? Or something along those lines.